inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java

```diff
@@ -252,36 +250,39 @@ public EvaluationResult evaluate(List<CAS> aCasses, DataSplitter aDataSplitter)
 
         // Train
         Trie<DictEntry> dict = createTrie();
+        List<Span> spans = new ArrayList<Span>();
         for (Sample sample : trainingSet) {
-            for (Span span : sample.getSpans()) {
-                learn(dict, span.getText(), span.getLabel());
-            }
+            spans.addAll(sample.getSpans());
+            
+        }
+        for (Span span : spans) {
+            learn(dict, span.getText(), span.getLabel());
         }
 
         // Predict
         List<LabelPair> labelPairs = new ArrayList<>();
+        List<TokenSpan> tokens = new ArrayList<TokenSpan>();
         for (Sample sample : testSet) {
-
-            for (TokenSpan token : sample.getTokens()) {
-                Trie<DictEntry>.Node node = dict.getNode(sample.getText(),
-                        token.getBegin() - sample.getBegin());
-                int begin = token.getBegin();
-                int end = token.getEnd();
-
-                String predictedLabel = NO_LABEL;
-                if (node != null && sample.hasTokenEndingAt(token.getBegin() + node.level)) {
-                    List<LabelStats> labelStats = node.value.getBest(1);
-                    if (!labelStats.isEmpty()) {
-                        predictedLabel = labelStats.get(0).getLabel();
-                    }
+            tokens.addAll(sample.getTokens());
+        }
+        for (TokenSpan token : tokens) {
+            Trie<DictEntry>.Node node = dict.getNode(sample.getText(),
+                    token.getBegin() - sample.getBegin());
+            int begin = token.getBegin();
+            int end = token.getEnd();
+            String predictedLabel = NO_LABEL;
+            if (node != null && sample.hasTokenEndingAt(token.getBegin() + node.level)) {
+                List<LabelStats> labelStats = node.value.getBest(1);
+                if (!labelStats.isEmpty()) {
+                    predictedLabel = labelStats.get(0).getLabel();
                 }
-                Optional<Span> coveringSpan = sample.getCoveringSpan(begin, end);
-                String goldLabel = NO_LABEL;
-                if (coveringSpan.isPresent()) {
-                    goldLabel = coveringSpan.get().getLabel();
-                }
-                labelPairs.add(new LabelPair(goldLabel, predictedLabel));
             }
+            Optional<Span> coveringSpan = sample.getCoveringSpan(begin, end);
```

Comment:
Hm, why is this now outside of the check `if (node !_ null && sample...)`? Not sure if this is correct.